### DISCUSSION TITLE
Example with c/c++

### DISCUSSION THREAD

**Discussion by huakim** [ksh93/ksh#749] [Q&A]
Can ksh be embeeded to the c/c++ code? This would be nice, if there are working examples, where embeeding of ksh implemented.

**Comment by HansH111** [comment]
I have since many years embedded the ksh2020 and the ksh93u+m in an application, but it requires some minor adaptions in main.c
I needed to adapt the sh_main extend the parameters for it to parse the ksh code which is already in memory....
I use a simple test program for testing.

```
#include <stdio.h>
#include <stdlib.h>
extern int sh_mainex(int, char *[], char *,char *,long);

int main(int argc, char *argv[], char *envp[])
{
long codelen=0;
char *code;

code=getenv("KSHCODE");
return sh_mainex(argc, argv, NULL,code,codelen);
}
```
Last time I compiled this was  with version Js 93u+m/1.0.8 as a static binary...

But maybe it can be done without adaptions, because I read...
        In addition, KSH-93 can be used as a library that can be embedded into an application to allow scripting.

Maybe someone can explain how to do this ?

**Reply by huakim** [reply]
but, there are no so much examples about how to read internal variables from korn shell session

**Comment by HansH111** [comment]
I had some time today, so made you an example.
The patch files works only for v1.0.8 of 2024-01-01
I am using Alpine Linux to compile and the current version in GitHub gives compile errors against the fortified headers.

```
Files in patch.zip:
SHOPT.sh.patch  enable only the things which are needed, so non interactive
defs.h.patch    removes errormsg when SCRIPTONLY=1
main.c.patch    makes changes to sh_main to accept also memory kshcode
testksh.c       very simple example for embedding ksh in a program
```

```
# Assume you have cloned the github repository.
$ git checkout v1.0.8
$ patch -u -b src/cmd/ksh93/SHOPT.sh       -i SHOPT.sh.patch
$ patch -u -b src/cmd/ksh93/include/defs.h -i defs.h.patch
$ patch -u -b src/cmd/ksh93/sh/main.c      -i main.c.patch

# compile ksh and libraries
$ bin/package make

# copy lib*.a libraries to local directory and compile example testksh
cp arch/*/lib/lib*.a .
gcc -s -static -o testksh testksh.c -L. libshell.a libdll.a libcmd.a libast.a

# run testksh example
$ export KSHCODE='ls -l testksh*;echo "hello world"; exit 1'
$ ./testksh
-rwxr-xr-x    1 root     root       1904840 May 15 15:34 testksh
-rwxr-xr-x    1 root     root           311 May 15 15:27 testksh.c
hello world
$ echo $?
1
```

Maybe there is an easier way to do this... 

[patch.zip](https://github.com/ksh93/ksh/files/15322462/patch.zip)

**Reply by HansH111** [reply]
After applying JohnoKing patch for Alpine fortified headers I adapted the patch.zip for the latest git version
So same instructions for patching and compiling

Results:
```
# run testksh example
$ export KSHCODE='ls -l testksh*;echo "hello world"; exit 1'
$ ./testksh
-rwxr-xr-x    1 root     root       1371576 May 19 11:42 testksh
-rwxr-xr-x    1 root     root           311 May 15 15:27 testksh.c
hello world
$ echo $?
1
```
No idea why the binary has almost lost 1/3 of its size...

[patch.zip](https://github.com/ksh93/ksh/files/15368915/patch.zip)
